phabricatoremails/settings.py

```diff
@@ -0,0 +1,122 @@
+from phabricatoremails.progress import StaticProgress, PhabricatorProgress
+from phabricatoremails.source import FileSource, PhabricatorSource
+from sqlalchemy import create_engine
+
+
+def _parse_logger(config: ConfigParser):
+    """Provides a logger set up according to our configuration
+
+    The regular logger is noisy and implements the MozLog JSON format, so we also
+    have a "dev" logger that simply prints out each log without any annotations.
+    The correct logger is chosen based on the provided configuration file.
+    """
+    if config.has_section("dev"):
+        return create_dev_logger()
+    else:
+        return create_logger()
+
+
+def _parse_source(config: ConfigParser, logger: Logger):
+    """Provides data-fetching implementations according to our configuration
+
+    To facilitate easier local development, we have a few different combinations of
+    ways to get data (Phabricator, or just a local file) as well as how we want
+    to perform at runtime (fetch-and-send emails once, or on a continuous loop).
+    Returns a "data source" and "progress" implementation
+    """
+    dev_source_file = config.get("dev", "file", fallback=None)
+    if dev_source_file:
+        return FileSource(pathlib.Path(dev_source_file).resolve()), StaticProgress(None)
+
+    host = config.get("phabricator", "host")
+    token = config.get("phabricator", "token")
+    poll_gap_seconds = int(config.get("phabricator", "poll_gap_seconds", fallback="60"))
+    override_since_key = config.get("dev", "since_key", fallback=None)
+    override_limit = config.get("dev", "limit", fallback=None)
+    if host.endswith("/"):
+        host = host[:-1]
+
+    source = PhabricatorSource(host, token, override_limit)
+    progress = (
+        StaticProgress(override_since_key)
+        if override_since_key
+        else PhabricatorProgress(logger, poll_gap_seconds)
+    )
+    return source, progress
+
+
+@dataclass
+class Settings:
+    """Stores and parses configuration the settings INI file
+
+    Some values are lazy-loaded from functions (instead of properties) because
+    they may have side-effects, such as connecting to an SMTP server.
+    """
+
+    source: Any
+    progress: Any
+    bugzilla_host: str
+    logger: Logger
+    sentry_dsn: str
+    _config: ConfigParser
+
+    def __init__(self, config: ConfigParser):
+        logger = _parse_logger(config)
+        source, progress = _parse_source(config, logger)
+        self.source = source
+        self.progress = progress
+        self.bugzilla_host = config.get("bugzilla", "host")
+        self.logger = logger
+        self.sentry_dsn = config.get("sentry", "dsn", fallback=None)
+        self._config = config
+
+    def db(self):
+        return Db(create_engine(self._config.get("db", "url")))
+
+    def mail(self):
+        from_address = self._config.get("email", "from_address")
+
+        if self._config.has_section("email-ses"):
```

Comment:
I think it's more robust to add a specific config option to set the email service (or any external service) rather than to solely rely on whether a config section is present or not. For example, what if someone accidentally adds both an `smtp` and `ses` section? I suggest adding configuration parameters for these types of settings to be more explicit.